src/lib/gadgets/bitwise.ts

```diff
@@ -0,0 +1,195 @@
+  );
+
+  // obtain pad length until the length is a multiple of 4*n for n-bit length lookup table
+  let padLength = length;
+  if (length % (4 * lengthXor) !== 0) {
+    padLength = length + 4 * lengthXor - (length % (4 * lengthXor));
+  }
+
+  // recursively build xor gadget chain
+  buildXor(a, b, outputXor, padLength, lengthXor);
+
+  // return the result of the xor operation
+  return outputXor;
+}
+
+// builds xor chain recursively
+function buildXor(
+  a: Field,
+  b: Field,
+  expectedOutput: Field,
+  padLength: number,
+  lengthXor: number
+) {
+  // if inputs are zero and length is zero, add the zero check
+  if (padLength === 0) {
+    Gates.zeroCheck(a, b, expectedOutput);
+
+    let zero = new Field(0);
+    zero.assertEquals(a);
+    zero.assertEquals(b);
+    zero.assertEquals(expectedOutput);
+  } else {
+    // nibble offsets
+    let first = lengthXor;
+    let second = first + lengthXor;
+    let third = second + lengthXor;
+    let fourth = third + lengthXor;
+
+    // slices of a
+    let in1_0 = witnessSlices(a, 0, first);
+    let in1_1 = witnessSlices(a, first, second);
+    let in1_2 = witnessSlices(a, second, third);
+    let in1_3 = witnessSlices(a, third, fourth);
+
+    // slices of b
+    let in2_0 = witnessSlices(b, 0, first);
+    let in2_1 = witnessSlices(b, first, second);
+    let in2_2 = witnessSlices(b, second, third);
+    let in2_3 = witnessSlices(b, third, fourth);
+
+    // slice of expected output
+    let out0 = witnessSlices(expectedOutput, 0, first);
+    let out1 = witnessSlices(expectedOutput, first, second);
+    let out2 = witnessSlices(expectedOutput, second, third);
+    let out3 = witnessSlices(expectedOutput, third, fourth);
+
+    // assert that the xor of the slices is correct, 16 bit at a time
+    Gates.xor(
+      a,
+      b,
+      expectedOutput,
+      in1_0,
+      in1_1,
+      in1_2,
+      in1_3,
+      in2_0,
+      in2_1,
+      in2_2,
+      in2_3,
+      out0,
+      out1,
+      out2,
+      out3
+    );
+
+    let nextIn1 = witnessNextValue(a, in1_0, in1_1, in1_2, in1_3, lengthXor);
+    let nextIn2 = witnessNextValue(b, in2_0, in2_1, in2_2, in2_3, lengthXor);
+    let nextExpectedOutput = witnessNextValue(
+      expectedOutput,
+      out0,
+      out1,
+      out2,
+      out3,
+      lengthXor
+    );
+
+    let next_length = padLength - 4 * lengthXor;
+    buildXor(nextIn1, nextIn2, nextExpectedOutput, next_length, lengthXor);
+  }
+}
+
+function assert(stmt: boolean, message?: string) {
+  if (!stmt) {
+    throw Error(message ?? 'Assertion failed');
+  }
+}
+
+function witnessSlices(f: Field, start: number, stop = -1) {
+  if (stop !== -1 && stop <= start)
+    throw Error('stop offset must be greater than start offset');
+
+  return Provable.witness(Field, () => {
+    let bits = f.toBits();
+    if (stop > bits.length) throw Error('stop must be less than bit-length');
+    if (stop === -1) stop = bits.length;
+
+    return Field.fromBits(bits.slice(start, stop));
```

Comment:
bitwise operations `>>` and `&` on `f.toBigint()` would be more efficient than slicing the bits.
also, since `f` is a variable, this probably goes through the provable `toBits` logic and snarky, which is always slow